be2-scala/src/main/scala/ch/epfl/pop/pubsub/PubSubMediator.scala

```diff
@@ -3,62 +3,125 @@ package ch.epfl.pop.pubsub
 import akka.event.LoggingReceive
 import akka.pattern.AskableActorRef
+import ch.epfl.pop.model.network.method.Broadcast
+import ch.epfl.pop.model.network.method.message.Message
+import ch.epfl.pop.model.network.{JsonRpcRequest, MethodType}
 import ch.epfl.pop.model.objects.Channel
-import ch.epfl.pop.pubsub.ClientActor.{SubscribeTo, UnsubscribeFrom}
-import ch.epfl.pop.pubsub.PubSubMediator.{SubscribeToAck, SubscribeToNAck, UnsubscribeFromAck, UnsubscribeFromNAck}
-import ch.epfl.pop.pubsub.graph.DbActor
+import ch.epfl.pop.pubsub.PubSubMediator._
+import ch.epfl.pop.pubsub.graph.DbActor.DbActorAck
+import ch.epfl.pop.pubsub.graph.validators.RpcValidator
+import ch.epfl.pop.pubsub.graph.{DbActor, GraphMessage}
 
 import scala.collection.mutable
 import scala.concurrent.ExecutionContext.Implicits.global
-import scala.concurrent._
+import scala.concurrent.{Await, Future}
 
-class PubSubMediator(val dbActor: AskableActorRef) extends Actor with ActorLogging with AskPatternConstants {
+class PubSubMediator extends Actor with ActorLogging with AskPatternConstants {
 
   // Map from channels to the collection of actors subscribed to the channel
   private var channelMap: mutable.Map[Channel, mutable.Set[ActorRef]] = mutable.Map.empty
+  lazy val dbActor: AskableActorRef = DbActor.getInstance
 
-  override def receive: Receive = LoggingReceive {
 
-    case SubscribeTo(channel) =>
-      val senderRef = sender // since we are asking dbActor, sender (client) will get overwritten!
-      (dbActor ? DbActor.ChannelExists(channel)).map {
-        case true => channelMap.get(channel) match {
-          case Some(set) =>
-            if (set.contains(senderRef)) {
-              val reason: String = s"Actor $senderRef was already subscribed to channel '$channel'"
-              log.info(reason)
-              senderRef ! SubscribeToNAck(channel, reason)
-            } else {
-              log.info(s"Subscribing $senderRef to channel '$channel'")
-              set += senderRef
-              senderRef ! SubscribeToAck(channel)
-            }
-          case _ =>
-            log.info(s"Subscribing $senderRef to channel '$channel'")
-            channelMap = channelMap + (channel -> mutable.Set(senderRef))
-            senderRef ! SubscribeToAck(channel)
-        }
+  private def subscribeTo(channel: Channel, clientActorRef: ActorRef): Future[PubSubMediatorMessage] = {
+    val ask: Future[PubSubMediatorMessage] = (dbActor ? DbActor.ChannelExists(channel)).map {
+      // if the channel exists in db, we can start thinking about subscribing a client to it
+      case DbActorAck() => channelMap.get(channel) match {
+        // if we have people already subscribed to said channel
+        case Some(set) =>
+          if (set.contains(clientActorRef)) {
+            val reason: String = s"Actor $clientActorRef was already subscribed to channel '$channel'"
+            log.info(reason)
+            SubscribeToNAck(channel, reason)
```

Comment:
I'm wondering, should this be a `NAck` ?
In practice the system could say "sure, you're subscribed" at this point.